### PR TITLE
Clear stale data using DeleteFilesInRanges

### PR DESCRIPTION
**Pull request by huachaohuang**
Optimize `clear_stale_data` by deleting files in multiple ranges at once.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -368,21 +368,26 @@ impl<T, C> Store<T, C> {
...
     fn clear_stale_data(&mut self) -> Result<()> {
         let t = Instant::now();
+
+        let mut ranges = Vec::new();
         let mut last_start_key = keys::data_key(b"");
         for region_id in self.region_ranges.values() {
             let region = self.region_peers[region_id].region();
             let start_key = keys::enc_start_key(region);
-            rocksdb::roughly_cleanup_range(&self.kv_engine, &last_start_key, &start_key)?;
+            ranges.push((last_start_key, start_key));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: How about if `last_start_key == start_key` for the first region?

File: src/util/rocksdb/mod.rs [hunk 2]
```diff
@@ -305,42 +320,15 @@ impl SliceTransform for NoopSliceTransform {
...
-    // Todo: use end_key directly after delete_files_in_range support [start, end) semantics.
-    let end = prev_key(end_key)?;
-    if start_key >= end.as_slice() {
-        return Ok(());
-    }
+pub fn roughly_cleanup_ranges(db: &DB, ranges: &[(Vec<u8>, Vec<u8>)]) -> Result<(), String> {
     for cf in db.cf_names() {
+        let mut delete_ranges = Vec::new();
+        for range in ranges {
+            assert!(range.0 <= range.1);
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/mod.rs]: I think we should skip it when range.start == range.end.

File: src/util/rocksdb/mod.rs [hunk 3]
```diff
@@ -305,42 +320,15 @@ impl SliceTransform for NoopSliceTransform {
...
+            delete_ranges.push(Range::new(&range.0, &range.1));
+        }
         let handle = get_cf_handle(db, cf)?;
-        db.delete_file_in_range_cf(handle, start_key, &end)?;
+        db.delete_files_in_ranges_cf(handle, &delete_ranges, /* include_end */ false)?;
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/mod.rs]: Please check if delete_ranges empty before involve delete_files_in_ranges_cf.
BTW, we only need to calculate ranges once and reuse it for each cf.

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -2729,4 +2748,219 @@ mod tests {
...
+        }
+
+        // Sleep 1s for compaction.
+        thread::sleep(Duration::from_secs(1));
+
+        db
+    }
+
+    #[test]
+    fn test_clear_stale_data() {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Can you test it in integration tests?

File: src/util/rocksdb/mod.rs [hunk 5]
```diff
@@ -305,42 +320,22 @@ impl SliceTransform for NoopSliceTransform {
...
-
-pub fn roughly_cleanup_range(db: &DB, start_key: &[u8], end_key: &[u8]) -> Result<(), String> {
-    if start_key > end_key {
-        return Err(format!(
-            "[roughly_cleanup_in_range] start_key({:?}) should't larger than end_key({:?}).",
-            start_key, end_key
-        ));
+pub fn roughly_cleanup_ranges(db: &DB, ranges: &[(Vec<u8>, Vec<u8>)]) -> Result<(), String> {
+    let mut delete_ranges = Vec::new();
+    for range in ranges {
```
Comment (BusyJay) [inline src/util/rocksdb/mod.rs]: Why not use `(ref start_key, ref end_key) in ranges`?

File: src/raftstore/store/store.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/store.rs]: There is a raftstore node in integration tests already. You can just new a single node cluster and do whatever you want.

**Comment by huachaohuang** [comment]
Working on the test.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

**Comment by huachaohuang** [comment]
OMG, all checks have passed, anyone seeing this please approve and merge this for me :)

**Pull request merged**